### DISCUSSION TITLE
Running custom code onDelete

### DISCUSSION THREAD

**Discussion by EvHaus** [prisma/prisma#15312] [Q&A] [answered]
In my app one of my Prisma models keeps track of files uploaded to AWS S3. I'm looking for some elegant way to ensure that the S3 assets are deleted from S3 when one of the records for this model in my DB is deleted. Sort of like a database trigger -- but where I can run my own Node.js code.

Is there anything inside of Prisma that can help me with this? Is there some hook I tap into that will get executed any time a record is deleted for a particular model?

**Comment by janpio** [comment, marked as answer]
If you explicitly call some form of `delete` operation on the model, you could probably use a middleware: https://www.prisma.io/docs/concepts/components/prisma-client/middleware If the deletion happens via a nested query, or even a referential action this might not be enough though unfortunately.

**Reply by EvHaus** [reply]
Looks really promising but unfortunately it doesn't trigger the middleware for a model if its parent is affected, so unfortunately it doesn't make my job much easier.

I'll go with a brute force approach then, but this might be a really nice feature to add to Prisma in the future somehow.

**Reply by janpio** [reply]
What exactly do you mean by "if its parent is affected"?

**Reply by EvHaus** [reply]
Sorry, I meant what you said earlier -- a referential action. For example, I have a data structure like this that represents a file browser (think MacOS's finder, or Windows' explorer):

- `Folder`
    - `File`
        - `Upload`

Each `Upload` model has an S3 asset associated with it that I want deleted anytime the `Upload` model, or its parent `File` or `Folder` models are deleted (via `onDelete` cascade).

Right now, the only way I can find to do this is to write manual logic anytime `Folder` or `File` or `Upload` are deleted to ensure it traverses the data tree and finds all related S3 assets. 

It would be great if Prisma could, for example, tell me that when I delete a `Folder` record, that will have a cascade delete on `Upload` records A, B and C -- and then I can issue a call to S3 to ensure those S3 assets get deleted appropriately.

**Reply by janpio** [reply]
If you defined your refernetial action in Prisma schema, Prisma could indeed be able to know that deleting a `Folder` will also delete specific `Upload` values. Would you expect this `onDelete` hook be called for these as well, based on the referential actions defined in the schema?

**Reply by EvHaus** [reply]
> Would you expect this `onDelete` hook be called for these as well, based on the referential actions defined in the schema?

Yes, exactly. That's what I'm looking for.

**Reply by janpio** [reply]
Would this be covered by https://github.com/prisma/prisma/issues/3547 or https://github.com/prisma/prisma/issues/9684? Maybe just missing a comment about thow it should work for referential actions?

**Reply by EvHaus** [reply]
Yeah, the hooks idea looks very promising!
